actioncable/app/javascript/action_cable/connection.js

```diff
@@ -0,0 +1,150 @@
+import ActionCable from "./index"
+import INTERNAL from "./internal"
+
+// Encapsulate the cable connection held by the consumer. This is an internal class not intended for direct user manipulation.
+
+const {message_types, protocols} = INTERNAL
+const supportedProtocols = protocols.slice(0, protocols.length - 1)
+
+const indexOf = [].indexOf
+
+class Connection {
+  constructor(consumer) {
+    this.open = this.open.bind(this)
+    this.consumer = consumer
+    this.subscriptions = this.consumer.subscriptions
+    this.monitor = new ActionCable.ConnectionMonitor(this)
+    this.disconnected = true
+  }
+
+  send(data) {
+    if (this.isOpen()) {
+      this.webSocket.send(JSON.stringify(data))
+      return true
+    } else {
+      return false
+    }
+  }
+
+  open() {
+    if (this.isActive()) {
+      ActionCable.log(`Attempted to open WebSocket, but existing socket is ${this.getState()}`)
+      return false
+    } else {
+      ActionCable.log(`Opening WebSocket, current state is ${this.getState()}, subprotocols: ${protocols}`)
+      if (this.webSocket) { this.uninstallEventHandlers() }
+      this.webSocket = new ActionCable.WebSocket(this.consumer.url, protocols)
+      this.installEventHandlers()
+      this.monitor.start()
+      return true
+    }
+  }
+
+  close({allowReconnect} = {allowReconnect: true}) {
+    if (!allowReconnect) { this.monitor.stop() }
+    if (this.isActive()) { return (this.webSocket ? this.webSocket.close() : undefined) }
+  }
+
+  reopen() {
+    ActionCable.log(`Reopening WebSocket, current state is ${this.getState()}`)
+    if (this.isActive()) {
+      try {
+        return this.close()
+      } catch (error) {
+        ActionCable.log("Failed to reopen WebSocket", error)
+      }
+      finally {
+        ActionCable.log(`Reopening WebSocket in ${this.constructor.reopenDelay}ms`)
+        setTimeout(this.open, this.constructor.reopenDelay)
+      }
+    } else {
+      return this.open()
+    }
+  }
+
+  getProtocol() {
+    return (this.webSocket ? this.webSocket.protocol : undefined)
+  }
+
+  isOpen() {
+    return this.isState("open")
+  }
+
+  isActive() {
+    return this.isState("open", "connecting")
+  }
+
+  // Private
+
+  isProtocolSupported() {
+    return indexOf.call(supportedProtocols, this.getProtocol()) >= 0
+  }
+
+  isState(...states) {
+    return indexOf.call(states, this.getState()) >= 0
+  }
+
+  getState() {
+    for (let state in WebSocket) { const value = WebSocket[state]; if (value === (this.webSocket ? this.webSocket.readyState : undefined)) { return state.toLowerCase() } }
+    return null
```

Comment:
This can be rewritten more naturally now. Don't need to preserve the single-line direct CoffeeScript translation as long as the result is the same.

Reply:
Yeah I don't like the translation of this line either. Is this what you had in mind:

```js
if (this.webSocket) {
  Object.keys(WebSocket).forEach(state => {
    if (WebSocket[state] === this.webSocket.readyState) {
      return state.toLowerCase()
    }
  })
}

return null
```
or something else?

Reply:
Do you like this better:

```js
if (this.webSocket) {
  for (let state in WebSocket) {
    if (WebSocket[state] === this.webSocket.readyState) {
      return state.toLowerCase()
    }
  })
}

return null
```

I also considered some alternatives, but they used methods that would require a polyfill for browsers that don't support ES2015 natively. That would be a new requirement for ActionCable users, so I wanted to avoid it.

Reply:
Yeah that's much better.

Reply:
Ok, I pushed that refactoring in https://github.com/rmacklin/rails/commit/dc5c137985943b7a39e99e02c78a7918e8a90833